File: backend/views/reservation_routes.py
from fastapi import APIRouter, HTTPException
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, validator
from datetime import date

from models.reservation import EmployeeReservation
from models.employee import Employee
from controllers.settings_controller import SettingsController

router = APIRouter(prefix="/api/employees/{employee_id}/reservations", tags=["reservations"])

# Pydantic models
class ReservationCreate(BaseModel):
    start_date: date
    end_date: date
    reserved_hours_per_day: float
    reason: Optional[str] = None
    
    @validator('end_date')
    def validate_dates(cls, v, values):
        if 'start_date' in values and v < values['start_date']:
            raise ValueError('End date must be after or equal to start date')
        return v
    
    @validator('reserved_hours_per_day')
    def validate_hours(cls, v):
        # Basic range check - more specific validation done in model/route handler
        if v < 0 or v > 24:
            raise ValueError('Reserved hours per day must be between 0 and 24')
        return v

class ReservationUpdate(BaseModel):
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    reserved_hours_per_day: Optional[float] = None
    reason: Optional[str] = None
    status: Optional[str] = None

@router.post("", response_model=Dict[str, Any])
async def create_reservation(employee_id: int, reservation: ReservationCreate):
    """Create a new reservation for an employee"""
    # Check if employee exists
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    try:
        new_reservation = EmployeeReservation.create(
            employee_id=employee_id,
            start_date=reservation.start_date,
            end_date=reservation.end_date,
            reserved_hours_per_day=reservation.reserved_hours_per_day,
            reason=reservation.reason if reservation.reason else None
        )
        return {'success': True, 'reservation': new_reservation.to_dict()}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to create reservation: {str(e)}")

@router.get("", response_model=List[Dict[str, Any]])
async def get_reservations(employee_id: int, include_cancelled: bool = False):
    """Get all reservations for an employee"""
    employee = Employee.get_by_id(employee_id)
    if not employee:
        raise HTTPException(status_code=404, detail="Employee not found")
    
    reservations = EmployeeReservation.get_by_employee(employee_id, include_cancelled)
    return [r.to_dict() for r in reservations]

@router.get("/{reservation_id}", response_model=Dict[str, Any])
async def get_reservation(employee_id: int, reservation_id: int):
    """Get a specific reservation"""
    reservation = EmployeeReservation.get_by_id(reservation_id)
    if not reservation:
        raise HTTPException(status_code=404, detail="Reservation not found")
    
    if reservation.employee_id != employee_id:
        raise HTTPException(status_code=403, detail="Reservation does not belong to this employee")
    
    return reservation.to_dict()

@router.put("/{reservation_id}", response_model=Dict[str, Any])
async def update_reservation(employee_id: int, reservation_id: int, update: ReservationUpdate):
    """Update a reservation"""
    reservation = EmployeeReservation.get_by_id(reservation_id)
    if not reservation:
        raise HTTPException(status_code=404, detail="Reservation not found")
    
    if reservation.employee_id != employee_id:
        raise HTTPException(status_code=403, detail="Reservation does not belong to this employee")
    
    # Filter out None values
    update_data = {k: v for k, v in update.dict().items() if v is not None}
    
    try:
        updated_reservation = reservation.update(**update_data)
        return {'success': True, 'reservation': updated_reservation.to_dict()}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update reservation: {str(e)}")

@router.delete("/{reservation_id}", response_model=Dict[str, Any])
async def delete_reservation(employee_id: int, reservation_id: int):
    """Delete a reservation"""
    reservation = EmployeeReservation.get_by_id(reservation_id)
    if not reservation:
        raise HTTPException(status_code=404, detail="Reservation not found")
    
    if reservation.employee_id != employee_id:
        raise HTTPException(status_code=403, detail="Reservation does not belong to this employee")
    
    try:
        reservation.delete()
        return {'success': True, 'message': 'Reservation deleted successfully'}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to delete reservation: {str(e)}")

@router.post("/{reservation_id}/cancel", response_model=Dict[str, Any])
async def cancel_reservation(employee_id: int, reservation_id: int):
    """Cancel a reservation"""
    reservation = EmployeeReservation.get_by_id(reservation_id)
    if not reservation:
        raise HTTPException(status_code=404, detail="Reservation not found")
    
    if reservation.employee_id != employee_id:
        raise HTTPException(status_code=403, detail="Reservation does not belong to this employee")
    
    try:
        cancelled_reservation = reservation.cancel()
        return {'success': True, 'reservation': cancelled_reservation.to_dict()}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to cancel reservation: {str(e)}")

File: backend-node/src/routes/reservationRoutes.js
const express = require('express');
const router = express.Router({ mergeParams: true }); // Important for accessing parent params
const EmployeeReservation = require('../models/reservation');
const Employee = require('../models/employee');

// Create reservation
router.post('/', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);

    // Check if employee exists
    const employee = Employee.getById(employeeId);
    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    const { start_date, end_date, reserved_hours_per_day, reason } = req.body;

    // Validate dates
    if (end_date < start_date) {
      return res.status(400).json({ detail: 'End date must be after or equal to start date' });
    }

    // Validate hours
    if (reserved_hours_per_day < 0 || reserved_hours_per_day > 24) {
      return res.status(400).json({ detail: 'Reserved hours per day must be between 0 and 24' });
    }

    const newReservation = EmployeeReservation.create(
      employeeId,
      start_date,
      end_date,
      reserved_hours_per_day,
      reason || null
    );

    res.json({ success: true, reservation: newReservation.toDict() });
  } catch (e) {
    if (e.message.includes('Reserved hours')) {
      return res.status(400).json({ detail: e.message });
    }
    res.status(500).json({ detail: `Failed to create reservation: ${e.message}` });
  }
});

// Get all reservations for an employee
router.get('/', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const includeCancelled = req.query.include_cancelled === 'true';

    const employee = Employee.getById(employeeId);
    if (!employee) {
      return res.status(404).json({ detail: 'Employee not found' });
    }

    const reservations = EmployeeReservation.getByEmployee(employeeId, includeCancelled);
    res.json(reservations.map(r => r.toDict()));
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Get specific reservation
router.get('/:reservationId', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const reservationId = parseInt(req.params.reservationId);

    const reservation = EmployeeReservation.getById(reservationId);
    if (!reservation) {
      return res.status(404).json({ detail: 'Reservation not found' });
    }

    if (reservation.employee_id !== employeeId) {
      return res.status(403).json({ detail: 'Reservation does not belong to this employee' });
    }

    res.json(reservation.toDict());
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Update reservation
router.put('/:reservationId', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const reservationId = parseInt(req.params.reservationId);

    const reservation = EmployeeReservation.getById(reservationId);
    if (!reservation) {
      return res.status(404).json({ detail: 'Reservation not found' });
    }

    if (reservation.employee_id !== employeeId) {
      return res.status(403).json({ detail: 'Reservation does not belong to this employee' });
    }

    const updatedReservation = reservation.update(req.body);
    res.json({ success: true, reservation: updatedReservation.toDict() });
  } catch (e) {
    if (e.message.includes('Reserved hours')) {
      return res.status(400).json({ detail: e.message });
    }
    res.status(500).json({ detail: `Failed to update reservation: ${e.message}` });
  }
});

// Delete reservation
router.delete('/:reservationId', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const reservationId = parseInt(req.params.reservationId);

    const reservation = EmployeeReservation.getById(reservationId);
    if (!reservation) {
      return res.status(404).json({ detail: 'Reservation not found' });
    }

    if (reservation.employee_id !== employeeId) {
      return res.status(403).json({ detail: 'Reservation does not belong to this employee' });
    }

    reservation.delete();
    res.json({ success: true, message: 'Reservation deleted successfully' });
  } catch (e) {
    res.status(500).json({ detail: `Failed to delete reservation: ${e.message}` });
  }
});

// Cancel reservation
router.post('/:reservationId/cancel', (req, res) => {
  try {
    const employeeId = parseInt(req.params.employeeId);
    const reservationId = parseInt(req.params.reservationId);

    const reservation = EmployeeReservation.getById(reservationId);
    if (!reservation) {
      return res.status(404).json({ detail: 'Reservation not found' });
    }

    if (reservation.employee_id !== employeeId) {
      return res.status(403).json({ detail: 'Reservation does not belong to this employee' });
    }

    const cancelledReservation = reservation.cancel();
    res.json({ success: true, reservation: cancelledReservation.toDict() });
  } catch (e) {
    res.status(500).json({ detail: `Failed to cancel reservation: ${e.message}` });
  }
});

module.exports = router;
